### PR TITLE
Fixes to http.send built-in

### PR DESCRIPTION
**Pull request by ashutosh-narkar**
Fixes to http.send to prevent unnecessary client creation
and to disable HTTP keep-alives. See commits for details.

Fixes: #4016
Fixes: #4015

### CODE REVIEW

File: topdown/http.go [hunk 1]
```diff
@@ -1188,23 +1195,19 @@ func newInterQueryCache(bctx BuiltinContext, key ast.Object, forceCacheParams *f
...
-		return nil, handleHTTPSendErr(c.bctx, err)
-	}
-
 	c.forceJSONDecode, err = getBoolValFromReqObj(c.key, ast.StringTerm("force_json_decode"))
 	if err != nil {
 		return nil, handleHTTPSendErr(c.bctx, err)
 	}
 
-	resp, err := checkHTTPSendInterQueryCache(c.bctx, c.key, c.httpReq, c.httpClient, c.forceJSONDecode, c.forceCacheParams)
+	resp, err := c.checkHTTPSendInterQueryCache(c.bctx, c.key, c.forceJSONDecode, c.forceCacheParams)
```
Comment (srenatus) [inline topdown/http.go:1203]: 🤔 Hrmm if this is method now has `c` as a receiver, what's the reason to pass `c.bctx`, `c.key`, etc?

**Review by srenatus — APPROVED** [review]
LGTM, one cleanup opportunity inside.

At a first glance, disabling keepalive seems odd -- it seems to be a thing we'd generally want. But that'll be a bigger feature, and this change should make things better 👍 

There a no tests here, I take it this is difficult to test properly in the golang unit test framework...?

**Comment by anderseknert** [comment]
I share @srenatus worries about disabling HTTP keep-alive. Seems like we're throwing out a a pretty fundamental performance enhancement for everyone in order to solve an edge case concerning a smaller group of users? Would setting `MaxIdleConns` on the transport not suffice?

```go
// MaxIdleConns controls the maximum number of idle (keep-alive)
// connections across all hosts. Zero means no limit.
MaxIdleConns int
```

**Comment by srenatus** [comment]
> I share @srenatus worries about disabling HTTP keep-alive. Seems like we're throwing out a a pretty fundamental performance enhancement for everyone in order to solve an edge case concerning a smaller group of users?

Oh I hadn't worried about that: we do no such thing. Right now, we create a new http.Client for each request, and each http.Client has its own Transport, and therefore there is *no actual connection re-use* right now. However, the Transport _by default_ attempts to *do connection re-use*, and keeps connections open. In our scenario, since we don't re-use the Transport that keeps the connection open, we're running out of file descriptors easily.

So, in my opinion this is the right fix for a real problem, and we're not sacrificing anything. I've mentioned the keepalive because in general it would be good to do that, to save the time needed to open a new connection, across multiple policy evals. But it's also tricky to get right, and _this PR_ is the proper solution to the problem at hand.

**Comment by anderseknert** [comment]
Gotcha! Thanks for expanding on that, Stephan 👍

File: topdown/http.go (hunk 1 above)
Comment (ashutosh-narkar) [reply topdown/http.go]: Updated. Removed the args to `c.checkHTTPSendInterQueryCache`.

**Comment by ashutosh-narkar** [comment]
> There a no tests here, I take it this is difficult to test properly in the golang unit test framework...?

I was trying out [this](https://pkg.go.dev/net/http/httptrace) package but did not find a good way to test this.

**Review by srenatus — APPROVED** [review]
Thanks

**Pull request merged**